sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala

```diff
@@ -369,4 +382,45 @@ trait ColumnResolutionHelper extends Logging {
       throws = true,
       allowOuter = allowOuter)
   }
+
+  private def resolveExpressionByPlanId(
+      e: Expression,
+      q: LogicalPlan): Expression = {
+    if (!e.exists(_.getTagValue(LogicalPlan.PLAN_ID_TAG).nonEmpty)) {
+      return e
+    }
+
+    e match {
+      case u: UnresolvedAttribute =>
+        resolveUnresolvedAttributeByPlanId(u, q).getOrElse(u)
+      case l: LeafExpression => l
+      case _ =>
+        e.mapChildren(c => resolveExpressionByPlanId(c, q))
+    }
+  }
+
+  private def resolveUnresolvedAttributeByPlanId(
+      u: UnresolvedAttribute,
+      q: LogicalPlan): Option[AttributeReference] = {
+    val planIdOpt = u.getTagValue(LogicalPlan.PLAN_ID_TAG)
+    if (planIdOpt.isEmpty) return None
+    val planId = planIdOpt.get
+    logDebug(s"Extract plan_id $planId from $u")
+
+    val planOpt = q.find(_.getTagValue(LogicalPlan.PLAN_ID_TAG).contains(planId))
+    if (planOpt.isEmpty) return None
+    val plan = planOpt.get
+    logDebug(s"Find child node $plan with plan_id==$planId")
+
+    try {
+      plan.resolve(u.nameParts, conf.resolver) match {
+        case Some(attr: AttributeReference) if plan.outputSet.contains(attr) => Some(attr)
```

Comment:
we don't need to do this check. Even if the attribute reference is dangling, we should still use it and fail later. You can check the behavior of normal dataframe using `df1.select(df2.col)`

Reply:
if we want to fail this case:
```
>>> df1 = spark.range(0, 10)
>>> df2 = spark.range(0, 10)
>>> df1
DataFrame[id: bigint]
>>> df1.select(df2.id)
DataFrame[id: bigint]
>>> df1.select(df2.id).show()
+---+                                                                           
| id|
+---+
|  0|
|  1|
|  2|
|  3|
|  4|
|  5|
|  6|
|  7|
|  8|
|  9|
+---+
```

I guess we should make it failed if can not find the matching node?

Reply:
yea, what's more
```
df1 = .... columns [a, b, c]
df2 = df1.select("a")
df2.select(df1.b)
```

this should fail as well, with missing attribute error.

Reply:
yes, add a new test for it.